linkerd/stack-metrics/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+pub struct Report<L: Hash + Eq> {
+    registry: Registry<L>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Layer {
+    metrics: Arc<Metrics>,
+}
+
+#[derive(Debug, Default)]
+struct Metrics {
+    create_total: Counter,
+    drop_total: Counter,
+    ready_total: Counter,
+    not_ready_total: Counter,
+    poll_millis: Counter,
+    error_total: Counter,
+}
+
+#[derive(Debug)]
+pub struct Service<S> {
+    inner: S,
+    metrics: Arc<Metrics>,
+    blocked_since: Option<Instant>,
+    // Helps determine when all instances are dropped.
+    _tracker: Arc<()>,
+}
+
+impl<L> NewLayer<L>
+where
+    L: Hash + Eq,
+{
+    pub fn new_layer(&self, labels: L) -> Layer {
+        let metrics = {
+            let mut registry = self.registry.lock().expect("stack metrics lock poisoned");
+            registry
+                .entry(labels.into())
+                .or_insert_with(|| Default::default())
+                .clone()
+        };
+        Layer { metrics }
+    }
+
+    pub fn report(&self) -> Report<L> {
+        Report {
+            registry: self.registry.clone(),
+        }
+    }
+}
+
+impl<L: Hash + Eq> Default for NewLayer<L> {
+    fn default() -> Self {
+        Self {
+            registry: Registry::default(),
+        }
+    }
+}
+
+impl<L: Hash + Eq> Clone for NewLayer<L> {
+    fn clone(&self) -> Self {
+        Self {
+            registry: self.registry.clone(),
+        }
+    }
+}
+
+impl<S> tower::layer::Layer<S> for Layer {
+    type Service = Service<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        self.metrics.create_total.incr();
+        Self::Service {
+            inner,
+            metrics: self.metrics.clone(),
+            blocked_since: None,
+            _tracker: Arc::new(()),
+        }
+    }
+}
+
+impl<T, S> tower::Service<T> for Service<S>
+where
+    S: tower::Service<T>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        match self.inner.poll_ready() {
+            Ok(Async::NotReady) => {
+                self.metrics.not_ready_total.incr();
+                if self.blocked_since.is_none() {
+                    self.blocked_since = Some(Instant::now());
+                }
+                Ok(Async::NotReady)
+            }
+            Ok(Async::Ready(())) => {
+                self.metrics.ready_total.incr();
+                if let Some(t0) = self.blocked_since.take() {
+                    let not_ready = Instant::now() - t0;
+                    self.metrics.poll_millis.add(not_ready.as_millis() as u64);
+                }
+                Ok(Async::Ready(()))
+            }
+            Err(e) => {
+                self.metrics.error_total.incr();
+                if let Some(t0) = self.blocked_since.take() {
+                    let not_ready = Instant::now() - t0;
+                    self.metrics.poll_millis.add(not_ready.as_millis() as u64);
+                }
```

Comment:
Any concerns about the performance implications of recording all these timestamps? It seems like if we are instrumenting every stack in the proxy with one of these layers, we are potentially adding a fair number of syscalls in the request path. 

It could be good to do some benchmarking and see if there's a noticeable perf impact from this? We might want to add a config to turn off the stack metrics, or something?

Reply:
How is this different from any of the other latency metrics we record?

Reply:
Furthermore, this is only triggred when a service is NotReady, which is fairly rare in the scheme of things. I'm not particularly worried about this? Unless you really think we should be?